src/entitytypes.js

```diff
@@ -976,6 +976,35 @@ SlugIdArray.prototype.indexOf = function(slug) {
   return -1;
 };
 
+/** Determines whether it includes a certain element, returning true or false as appropriate. */
+SlugIdArray.prototype.includes = function(slug) {
+  return this.buffer.includes(slugIdToBuffer(slug));
+};
+
+/** The shift() method removes element. */
+SlugIdArray.prototype.shift = function() {
+  if (this.length === 0) {
+    return;
+  }
+
+  this.buffer.copy(this.buffer, 0, SLUGID_SIZE);
+  this.avail  += 1;
+  this.length -= 1;
+  this.realloc();
+};
+
+/** The pop() method removes the last element. */
+SlugIdArray.prototype.pop = function() {
+  if (this.length === 0) {
+    return;
+  }
+
+  this.buffer.copy(this.buffer, (this.length - 1) * SLUGID_SIZE, this.length * SLUGID_SIZE);
```

Comment:
Do we need to do this? realloc is designed to allocated more memory than necessary, and less if too much is allocated... 

Why would we copy the buffer?

---
If you open an advanced algorithms book and lookup "amortized analysis"  and/or dynamic arrays, you'll find that by doubling the array size every time we recallocate `push()` has an amortized complexity of `O(1)`. Same for `pop()` if we only reallocate the array when less than a third of the array is used...
Choice of "doubling in size" and reduce size if "less than a third is used", can be adjusted as long as both of them are fractions and are two different fractions, then it's not possible to make a sequence of push/pop operations that constantly reallocate the array up and down..

https://en.wikipedia.org/wiki/Dynamic_array
https://en.wikipedia.org/wiki/Amortized_analysis

Reply:
TIL about amortized analysis. Thanks @jonasfj. Right, we don't need to copy the buffer.